service/controller/resource/subnet/resource.go

```diff
@@ -367,41 +379,44 @@ func isSubnetAllowedToStorageAccount(ctx context.Context, storageAccount storage
 }
 
 // This functions decides whether or not the ARM deployment is out of date.
-// For that, we use the Generation field from the AzureCluster CR. This Generation field should change when there is a change in the CR.
-func (r *Resource) isDeploymentOutOfDate(ctx context.Context, cr capzv1alpha3.AzureCluster, currentDeployment azureresource.DeploymentExtended) (bool, error) {
-	crVersion := strconv.FormatInt(cr.ObjectMeta.Generation, 10)
+// We only take into consideration the subnet's name and CIDR.
+func (r *Resource) isDeploymentOutOfDate(ctx context.Context, allocatedSubnet *capzv1alpha3.SubnetSpec, currentDeployment azureresource.DeploymentExtended) (bool, error) {
 	currentParams, ok := currentDeployment.Properties.Parameters.(map[string]interface{})
 	if !ok {
 		return false, microerror.Maskf(wrongTypeError, "expected '%T', got '%T'", map[string]interface{}{}, currentDeployment.Properties.Parameters)
 	}
 
-	deploymentVersion, ok := currentParams["azureClusterVersion"].(map[string]interface{})["value"].(string)
+	subnetCidr, ok := currentParams["subnetCidr"].(map[string]interface{})["value"].(string)
+	if !ok {
+		return false, microerror.Maskf(wrongTypeError, "expected 'string', got '%T'", currentParams["subnetCidr"].(map[string]interface{})["value"])
+	}
+
+	nodepoolName, ok := currentParams["nodepoolName"].(map[string]interface{})["value"].(string)
 	if !ok {
-		return false, microerror.Maskf(wrongTypeError, "expected 'string', got '%T'", currentDeployment.Properties.Parameters)
+		return false, microerror.Maskf(wrongTypeError, "expected 'string', got '%T'", currentParams["nodepoolName"].(map[string]interface{})["value"])
 	}
 
-	r.logger.LogCtx(ctx, "level", "debug", "message", "Checking if deployment is out of date", "azureClusterVersion", crVersion, "deploymentParameter", deploymentVersion)
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("Checking if deployment is out of date for %#q", nodepoolName), "desiredSubnetName", allocatedSubnet.Name, "deploymentSubnetName", nodepoolName, "desiredSubnetCidr", allocatedSubnet.CIDRBlocks[0], "deploymentSubnetCidr", subnetCidr)
 
-	return crVersion != deploymentVersion, nil
+	return allocatedSubnet.Name != nodepoolName || allocatedSubnet.CIDRBlocks[0] != subnetCidr, nil
```

Comment:
This is the main change that makes the deployment to be submitted only when any of the two fields change.